com.unity.netcode.gameobjects/Runtime/Serialization/FastBufferWriter.cs

```diff
@@ -0,0 +1,772 @@
+
+namespace Unity.Netcode
+{
+    public struct FastBufferWriter : IDisposable
+    {
+        internal unsafe byte* BufferPointer;
+        internal int PositionInternal;
+        private int m_Length;
+        internal int CapacityInternal;
+        internal readonly int MaxCapacityInternal;
+        private readonly Allocator m_Allocator;
+#if DEVELOPMENT_BUILD || UNITY_EDITOR
+        internal int AllowedWriteMark;
+        private bool m_InBitwiseContext;
+#endif
+
+        /// <summary>
+        /// The current write position
+        /// </summary>
+        public int Position
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => PositionInternal;
+        }
+
+        /// <summary>
+        /// The current total buffer size
+        /// </summary>
+        public int Capacity
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => CapacityInternal;
+        }
+
+        /// <summary>
+        /// The maximum possible total buffer size
+        /// </summary>
+        public int MaxCapacity
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => MaxCapacityInternal;
+        }
+
+        /// <summary>
+        /// The total amount of bytes that have been written to the stream
+        /// </summary>
+        public int Length
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => PositionInternal > m_Length ? PositionInternal : m_Length;
+        }
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void CommitBitwiseWrites(int amount)
+        {
+            PositionInternal += amount;
+#if DEVELOPMENT_BUILD || UNITY_EDITOR
+            m_InBitwiseContext = false;
+#endif
+        }
+
+        /// <summary>
+        /// Create a FastBufferWriter.
+        /// </summary>
+        /// <param name="size">Size of the buffer to create</param>
+        /// <param name="allocator">Allocator to use in creating it</param>
+        /// <param name="maxSize">Maximum size the buffer can grow to. If less than size, buffer cannot grow.</param>
+        public unsafe FastBufferWriter(int size, Allocator allocator, int maxSize = -1)
+        {
+            void* buffer = UnsafeUtility.Malloc(size, UnsafeUtility.AlignOf<byte>(), allocator);
+#if DEVELOPMENT_BUILD || UNITY_EDITOR
+            UnsafeUtility.MemSet(buffer, 0, size);
+#endif
+            BufferPointer = (byte*)buffer;
+            PositionInternal = 0;
+            m_Length = 0;
+            CapacityInternal = size;
+            m_Allocator = allocator;
+            MaxCapacityInternal = maxSize < size ? size : maxSize;
+#if DEVELOPMENT_BUILD || UNITY_EDITOR
+            AllowedWriteMark = 0;
+            m_InBitwiseContext = false;
+#endif
+        }
+
+        /// <summary>
+        /// Frees the allocated buffer
+        /// </summary>
+        public unsafe void Dispose()
```

Comment:
when and where you'd imagine this `Dispose()` method to be called?
I'm not seeing a destructor/finalizer calling this and wondering if this would mean we're pushing memory management onto users' plate?

Reply:
It'll be called by the messaging system whenever a buffer is no longer needed. I implemented it as IDisposable to support generic use of the serializer outside the messaging system since tools team has already requested the ability to do that for, i.e., file IO.

Reply:
my worry is that we're not applying [idiomatic disposing pattern](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose) here.
we should call `Dispose()` from the destructor/finalizer and also implement `Dispose(bool disposing)` methods to call them delicately as `Dispose(false)` and `Dispose(true)` in the right contexts.

Reply:
Structs can't have finalizers/destructors. Only classes can have those.